src/ScriptEditor/ui/ScriptEditorRoot.tsx

```diff
@@ -444,6 +436,19 @@ function Root(props: IProps): React.ReactElement {
     onOpenPreviousTab,
   });
 
+  const onTabClickRef = useRef(onTabClick);
+  onTabClickRef.current = onTabClick;
+  const parseCodeRef = useRef(parseCode);
+  parseCodeRef.current = parseCode;
```

Comment:
and this is also one of the ugliest things I've done this year xD

Reply:
That was a quick and dirty solution. I'm looking for something better here.

Reply:
Is the entire point of these refs to allow you to pass the lint rule while not passing dependencies to the useEffect?

If so, two other possible solutions, both also ugly in their own way:
1. Just use an eslint ignore next line comment instead of making runtime changes
2. Default parameters in the useEffect function, e.g.

```js
useEffect((parseCode = parseCode, onTabClick = onTabClick) => {
    if (currentScript !== null) {
      const tabIndex = currentTabIndex();
      if (typeof tabIndex === "number") onTabClick(tabIndex);
      parseCode(currentScript.code);
    }
  }, []);
```

Reply:
I actually think there was a bug in this code. As far as I remember, these custom vim commands did not work, probably because of wrong hook deps. I need ro test it.

These refs allow you not to pass dependensies, but still use the latest values from the last rerender. They don't trigger effects or memoized hook recalculations, they solve both linter and runtime issues... but they look ugly xD

Reply:
> 2\. Default parameters in the useEffect function

That's something I haven't seen before.

Reply:
> These refs allow you not to pass dependensies, but still use the latest values from the last rerender. They don't trigger effects or memoized hook recalculations, they solve both linter and runtime issues... but they look ugly xD

The effect is only firing exactly once on the first render of the component, the effect doesn't happen again after that. So it's not using "the latest value from the last rerender", it's just using the value from the first render, because that effect never gets called again.

> (default parameters) That's something I haven't seen before. And I don't think it's even supported by React as effects have no any arguments.

It works because useEffects don't receive any arguments, which leads to the default value always being used. The default values would just be the values of those functions from the first render (so functionally identical to what it is now)

I got the idea essentially from here: https://stackoverflow.com/questions/55840294/how-to-fix-missing-dependency-warning-when-using-useeffect-react-hook/61183162#61183162

Reply:
Explanation: refs work like class fields in the class-based react components. They are the same objects which are passed to every render call in the component (`current` can be different, but `ref` says the same).

So if during the effect call, which happens only on component mount, you register some event handlers or some delayed action on timeout/interval, etc... using a ref - you will use the updated ref.current value whenever the handler you registered is actually triggered. But if you pass down a regular constant, not ref, then you will basically use a fixed value which will never get changed, unless you put this value as dependency and re-register you handler with a new value.

Reply:
1. Each of the 3 buttons does increase the actual counter, they are just console.logging stale values.
2. Your example code is not like the actual game code here.
    * Your code is running the ref'd function at a time other than immediately (by placing it as an onclick). This provides a possibility for the function to become "stale" and need to get an updated reference. Game code in this useEffect is running immediately in the useRef and never again: there is no possibility for staleness.
    * Game code is referencing a function which is defined inside the same component, which means just directly referencing that function is already an "up to date reference" to the function even if the useEffect was ran more than once. Although if the useEffect was adding delayed code (which it's not) a ref might still make sense even with a function defined in the same component.

Reply:
> 1. Each of the 3 buttons does increase the actual counter, they are just console.logging stale values.

But that's exactly the point! Any non-memoized value (all except useCallback, useMemo, setState, useRef and static constants defined outside of component) you pass to useEffect/useCallback/useMemo without mentioning it as dependency will be "stale".

> 2. Your example code is not like the actual game code here.

Yes, but only in case of this specific place cause it does not have any kind of delays (timeouts/intervals, microtasks) or subscriptions (e.g. event listeners) and is executed immediately after the first render and only once.

My example just makes it obvious that without using ref (or passing down value to dependencies list) `useEffect` won't work as you expect it to in cases of delayed actions. As an example, you can check this code https://github.com/bezrodnov/bitburner-src/blob/enforce-eslint-rules/src/ScriptEditor/ui/useVimEditor.tsx#L27 where without using ref the code will not work properly as it will use stale values instead of the actual ones.


As for passing down arguments to useEffect's effect function - this is a very hacky thing that I have never seen since React hooks release. Not only it looks strange, but also does not solve the issue as react hooks plugin still complains about these arguments not being passed as hook dependencies:
![image](https://github.com/bitburner-official/bitburner-src/assets/5138959/9e7ad5b8-4cab-47b6-8897-1029518e910a)

Reply:
> Just use an eslint ignore next line comment instead of making runtime changes

This is also something I try to avoid at all costs. It's just not safe at the long run unless this code never changes again or whoever changes it in the future understands and notices that there's a lint rule being ignored here and knows how to fix it.

Reply:
> useEffect was needed for vim
That is because vim needs to be able to call fresh versions of the functions later on, when that functionality is invoked through vim keybindings in a delayed manner. This instance is not like that, there is no delayed manner in which the functions are potentially being used inside useEffect code, it's all immediate.

> default parameters
Yeah this doesn't even seem to work for clearing the lint failure, so it's off the table. Based on what I saw in that stackoverflow link I thought it was enough to satisfy the lint rule, my bad.


I get the aversion to using a lint ignore statement. If you want to keep the useRefs, at least get rid of the updating of ref.current every render - there is no point updating a ref when we know it is never going to be used again (although maybe this would also demand a variable rename, so it's clear that these refs are not kept current)

Reply:
So, I've been following along here... It feels like this shouldn't be a separate effect at all? Can't it be put in beforeMount or onMount?

Reply:
> Can't it be put in beforeMount or onMount?

This is a type of change I tried to avoid in this PR cause it might introduce bugs. This PR was purely about enforcing lint rules without changing behavior (but I realize that in some places I've already broken this intention).
For now, I just added a lint ignore instruction.